nixos/modules/services/web-apps/akkoma.nix

```diff
@@ -0,0 +1,1031 @@
+      description = "Akkoma social network configuration";
+      reloadTriggers = [ configFile ] ++ secretPaths;
+
+      unitConfig.PropagatesReloadTo = [ "akkoma.service" ];
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        UMask = "0077";
+
+        RuntimeDirectory = "akkoma";
+        RuntimeDirectoryMode = "0711";
+
+        ExecStart =
+          (if cfg.dist.cookie == null
+            then [ genScript ]
+            else [ copyScript ])
+          ++ optionals cfg.initSecrets [ initSecretsScript ]
+          ++ [ configScript ];
+        ExecReload = optionals cfg.initSecrets [ initSecretsScript ]
+          ++ [ configScript ];
+      };
+    };
+
+    systemd.services.akkoma-initdb = mkIf cfg.initDb.enable {
+      description = "Akkoma social network database setup";
+      requires = [ "akkoma-config.service" ];
+      requiredBy = [ "akkoma.service" ];
+      after = [ "akkoma-config.service" "postgresql.service" ];
+      before = [ "akkoma.service" ];
+
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        UMask = "0077";
+        ExecStart = initDbScript;
+        PrivateTmp = true;
+      } // optionalAttrs (db ? socket_dir || db ? socket) {
+        User = cfg.initDb.username;
+      };
+    };
+
+    systemd.services.akkoma = let
+      runtimeInputs = with pkgs; [ coreutils gawk gnused ] ++ cfg.extraPackages;
+    in {
+      description = "Akkoma social network";
+      documentation = [ "https://docs.akkoma.dev/stable/" ];
+
+      # This service depends on network-online.target and is sequenced after
+      # it because it requires access to the Internet to function properly.
+      bindsTo = [ "akkoma-config.service" ];
+      wants = [ "network-online.service" ];
+      wantedBy = [ "multi-user.target" ];
+      after = [
+        "akkoma-config.target"
+        "network.target"
+        "network-online.target"
+        "postgresql.service"
+      ];
+
+      confinement.packages = runtimeInputs;
+      path = runtimeInputs;
+
+      serviceConfig = {
+        Type = "exec";
+        User = cfg.user;
+        Group = cfg.group;
+        UMask = "0077";
+
+        RuntimeDirectory = "akkoma";
+        RuntimeDirectoryMode = "0711";
+        RuntimeDirectoryPreserve = true;
+        StateDirectory = "akkoma akkoma/static akkoma/uploads";
+        StateDirectoryMode = "0700";
+
+        BindReadOnlyPaths = [ "/etc/hosts" "/etc/resolv.conf" ]
+          ++ optionals (db ? socket_dir)
+            [ "${db.socket_dir}:${db.socket_dir}:norbind" ]
+          ++ optionals (db ? socket)
+            [ "${db.socket}:${db.socket}:norbind" ]
+          ++ optionals (cfg.termsOfService != null)
+            [ "${cfg.termsOfService}:${staticDir}/static/terms-of-service.html:norbind" ]
+          ++ mapAttrsToList (key: val:
+              "${val.package}:${staticDir}/frontends/${val.name}/${val.ref}:norbind")
+            cfg.frontends;
```

Comment:
It seems that everyone and their brother in nixpkgs adds `/nix/store` to BindReadOnlyPaths. I think _not_ doing so will be fragile, because Nix packages can always start referencing some other package's file by a symlink.

Reply:
The confinement module should take care of that automatically, at least for executable dependencies.

Reply:
Wouldn't it make sense to `mkIf confined` all of these settings that have no effect otherwise?

Reply:
The attribute is now constructed with `mkMerge` and `mkIf`, making the extra bind mounts optional.